File: src/main/java/algo/breadthFirstSearch/SlidingPuzzle_773.java
package algo.breadthFirstSearch;

import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Map;

public class SlidingPuzzle_773 {
  private static final String TARGET_STATE = "123450";
  private static final Map<Integer, int[]> DIRECTIONS = new HashMap<>()
  {
    {
      put(0, new int[]{1, 3});
      put(1, new int[]{0, 2, 4});
      put(2, new int[]{1, 5});
      put(3, new int[]{0, 4});
      put(4, new int[]{1, 3, 5});
      put(5, new int[]{2, 4});
    }
  };

  public static int slidingPuzzle(int[][] board) {
    return bfs(board, 0);
  }

  private static int bfs(int[][] state, int steps) {
    var states = new LinkedList<String>();
    var visited = new HashSet<String>();
    var initial = serialize(state);
    states.offer(initial);
    visited.add(initial);
    while (!states.isEmpty()) {
      var size = states.size();
      while (size-- > 0) {
        var st = states.poll();
        if (st.equals(TARGET_STATE)) {
          return steps;
        }
        var pos = st.indexOf('0');
        var dir = DIRECTIONS.get(pos);
        for (var d : dir) {
          var tmp = st.charAt(d);
          var next = st.replace('0', '6').replace(tmp, '0').replace('6', tmp);
          if (visited.add(next)) {
            states.offer(next);
          }
        }
      }
      steps++;
    }
    return -1;
  }

  private static String serialize(int[][] state) {
    var ser = "";
    for (var s : state) {
      for (var v : s) {
        ser += v;
      }
    }
    return ser;
  }

  public static void main(String[] args) {
    System.out.println(slidingPuzzle(new int[][]{{1,2,3},{4,0,5}}));
    System.out.println(slidingPuzzle(new int[][]{{1,2,3},{5,4,0}}));
    System.out.println(slidingPuzzle(new int[][]{{4,1,2},{5,0,3}}));
  }
}

File: src/main/kotlin/algo/breadthFirstSearch/SlidingPuzzle_773.kt
package algo.breadthFirstSearch

import java.util.LinkedList
import kotlin.text.StringBuilder

class KtSlidingPuzzle_773 {
    companion object {
        const val TARGET = "123450"
        // next movements when empty cell at index 0 to 5
        val direction = mapOf(
            0 to listOf(1, 3),
            1 to listOf(0, 2, 4),
            2 to listOf(1, 5),
            3 to listOf(0, 4),
            4 to listOf(1, 3, 5),
            5 to listOf(2, 4)
        )
    }
    fun slidingPuzzle(board: Array<IntArray>): Int {
        fun nextStates(position: Int, currentState: String): List<String> {
            val states: MutableList<String> = mutableListOf()
            direction[position]?.forEach {
                val tmp = currentState[it]
                val sb = StringBuilder(currentState)
                sb[it] = sb[position]
                sb[position] = tmp
                states.add(sb.toString())
            }
            return states
        }
        val state = board.flatMap { it.asIterable() }.joinToString("")
        val visited: MutableSet<String> = HashSet()
        var steps = 0
        val q = LinkedList<String>()
        q.offer(state)
        while (q.isNotEmpty()) {
            val size = q.size
            for (i in 1..size) {
                val st = q.poll()
                if (st == TARGET) {
                    return steps
                }
                if (!visited.add(st)) {
                    continue
                }
                nextStates(st.indexOf('0'), st).forEach { q.offer(it) }
            }
            steps++
        }
        return -1
    }
}

fun main() {
    val puzzle = KtSlidingPuzzle_773()
    println(puzzle.slidingPuzzle(arrayOf(intArrayOf(1,2,3), intArrayOf(4,0,5))))
    println(puzzle.slidingPuzzle(arrayOf(intArrayOf(1,2,3), intArrayOf(5,4,0))))
    println(puzzle.slidingPuzzle(arrayOf(intArrayOf(4,1,2), intArrayOf(5,0,3))))
}